dot4gravity/src/lib.rs

```diff
@@ -1,8 +1,707 @@
+const NUM_OF_BLOCKS: usize = 10;
+
+type Player = u32;
+
+/// Represents a cell of the board.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+enum Cell {
+    Empty,
+    Bomb([Option<Player>; NUM_OF_PLAYERS]),
+    Block,
+    Stone(Player),
+}
+
+impl Cell {
+    /// Tells if a cell is suitable for dropping a bomb.
+    fn is_valid_for_dropping_bomb(&self) -> bool {
+        match self {
+            Cell::Empty => true,
+            Cell::Bomb([Some(_), None]) => true,
+            _ => false,
+        }
+    }
+
+    /// Tells if a stone can traverse the cell.
+    fn is_travesable(&self) -> bool {
+        *self == Cell::Empty
+    }
+
+    /// Tells if a cell is of type 'bomb'
+    fn is_bomb(&self) -> bool {
+        match self {
+            Cell::Bomb([Some(_), None]) => true,
+            Cell::Bomb([Some(_), Some(_)]) => true,
+            _ => false,
+        }
+    }
+
+    /// Tells if a cell must be cleared when it's affected by an explosion.
+    fn is_explodable(&self) -> bool {
+        *self != Cell::Block
+    }
+}
+
+/// Coordinates for a cell in the board.
+#[derive(Clone, Copy, Debug, PartialEq)]
+struct Coordinates {
+    row: usize,
+    col: usize,
+}
+
+impl Coordinates {
+    /// Tells if a cell is inside the board.
+    fn is_inside_board(&self) -> bool {
+        self.row < BOARD_WIDTH && self.col < BOARD_HEIGHT
+    }
+
+    /// Tells if a cell is in the oposite of a side.
+    fn is_opposite_cell(&self, side: Side) -> bool {
+        match side {
+            Side::North => self.row == BOARD_HEIGHT - 1,
+            Side::East => self.col == 0,
+            Side::South => self.row == 0,
+            Side::West => self.col == BOARD_WIDTH - 1,
+        }
+    }
+}
+
+/// Sides of the board from which a player can drop a stone.
+#[derive(Clone, Debug)]
+enum Side {
+    North,
+    East,
+    South,
+    West,
+}
+
+#[derive(Clone, Debug)]
+struct Board {
+    cells: [[Cell; BOARD_WIDTH]; BOARD_HEIGHT],
+}
+
+impl Board {
+    pub fn new() -> Board {
+        Board {
+            cells: [[Cell::Empty; BOARD_WIDTH]; BOARD_HEIGHT],
+        }
+    }
+
+    fn get_cell(&self, position: &Coordinates) -> Cell {
+        self.cells[position.row][position.col]
+    }
+
+    fn change_cell(&mut self, position: Coordinates, cell: Cell) {
+        self.cells[position.row][position.col] = cell;
+        assert_eq!(self.cells[position.row][position.col], cell);
+    }
+
+    fn populate_with_random_blocks(mut board: Board, mut num_of_blocks: usize) -> Board {
+        let mut blocks = Vec::<Coordinates>::new();
+        let mut rng = rand::thread_rng();
+
+        while num_of_blocks > 0 {
+            let block_coordinates = Coordinates {
+                row: rng.gen_range(0..BOARD_HEIGHT),
+                col: rng.gen_range(0..BOARD_WIDTH),
+            };
+            if !blocks.contains(&block_coordinates) {
+                num_of_blocks -= 1;
+                blocks.push(block_coordinates)
+            }
+        }
```

Comment:
This could iterate more than we need - can we take random samples from two vectors? `choose_multiple` might work nicely here.

Reply:
I agree. changed!